DM-44503: Clean up Butler server REST API

dhirving opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Use query parameters when getting dataset type

Dataset type names may include slashes, and FastAPI's router does not correctly handle slashes in path fragments even if they are encoded.

dhirving committed:
Add wrapper object for get_dataset_type response

Add an outer object to the get_dataset_type JSON format to allow for backwards-compatible addition of properties in the future.

dhirving committed:
Declare a Pydantic model for dimension universe response

Instead of declaring that the server's dimension universe response type is 'dict', specify which Pydantic model it is using.  This aids maintainability by making it explicit that you can't change this serialized format without breaking backwards compatibility.

dhirving committed:
Ensure UUID is in correct format before sending to server

Since the ID in get_dataset is being encoded into the path, make sure it is in the expected format to avoid potential encoding shenanigans.

dhirving committed:
Remove datastore_records from get_dataset REST API

datastore_records is a DirectButler-only concept that we have never been able to return from Butler server, so remove it from the REST API.

dhirving committed:
Add wrapper object for get_dataset

Add an outer object to the get_dataset response JSON to allow for backwards-compatible addition of properties in the future.

dhirving committed:
Rename dataset_type_name for consistency

dhirving committed:
Move find_dataset dataset type to POST body

Dataset type names can include slashes, which can not be used in path variables in FastAPI.  There was also no real reason this variable should be passed in the path instead of in the body with the rest of the query.

dhirving committed:
Remove datastore records from find_dataset REST API

datastore_records is a DirectButler-only concept that we have never been able to return from Butler server, so remove it from the REST API.

dhirving committed:
Add a docstring

andy-slac commented on python/lsst/daf/butler/dimensions/_config.py:
```diff
@@ -156,6 +156,32 @@ def _updateWithOtherConfigFile(self, file: Config | str | ResourcePath | Mapping
+        """
+        return SerializedDimensionConfig.model_validate(self.toDict())
+
+    @staticmethod
+    def from_simple(simple: SerializedDimensionConfig) -> DimensionConfig:
```
I think in other places we prefer to use `@classmethod` for this so you can do `cls(...)` to support potential subclassing. Not sure it matters here though.

dhirving replied:
Generally speaking I'm not a big fan of that pattern, since it assumes subclasses have the same constructor signature as the base class.

And yeah specifically here no subclasses exist or should exist (and if they did, they would have a different serialized form with their own data so the function signatures wouldn't match up.)

andy-slac commented on python/lsst/daf/butler/dimensions/_config.py:
```diff
@@ -156,6 +156,32 @@ def _updateWithOtherConfigFile(self, file: Config | str | ResourcePath | Mapping
+
+        Returns
+        -------
+        model : `SerializedUniverseConfig`
```
```suggestion
        model : `SerializedDimensionConfig`
```

andy-slac reviewed (approved):
Looks great, a couple of minor comments.

timj commented on python/lsst/daf/butler/remote_butler/server/handlers/_external.py:
```diff
@@ -83,76 +85,60 @@ async def get_client_config() -> dict[str, Any]:
 
 @external_router.get(
-    "/v1/dataset_type/{dataset_type_name}",
+    "/v1/dataset_type",
```
If we know that we are enforcing that a name has no slashes it would be really nice from an API perspective if we did allow this simple form for retrieving the dataset type information. It seems like a very natural interface for single dataset type with no wildcards returning the details.

dhirving committed:
Fix incorrect type name in documentation

Co-authored-by: Andy Salnikov <salnikov@slac.stanford.edu>

dhirving committed:
Put dataset type parameter in path instead of query string

Tim likes it better this way.

dhirving committed:
Format UUIDs consistently in path fragments

Use the same logic for formatting UUIDs in URL path fragments everywhere we do it.
